### PR TITLE
[BUG] - bugfix | MSTL does not support a trend_forecaster that needs `fh` during `fit`

### PR DESCRIPTION
**Pull request by Abelarm**
#### Reference Issues/PRs
#5666 

#### What does this implement/fix? Explain your changes.
It enables to use estimators that needs `fh` as `trend_forecaster` for `MSTL` as shown in the issue.

##### Implementation details
During the fit of the `MSTL`(`_GeneralisedStatsForecastAdapter`) I add to the `trend_forecaster` object a property on the fly called `_fh` (with value coming from `fh` of it, only if it's not null).

The inside the fit of `StatsForecastBackAdapter` I get the property (None if not present) ad pass it to the `estimator.fit`

I am not sure if this is the most elegant solution...

#### Does your contribution introduce a new dependency? If yes, which one?
No

#### What should a reviewer concentrate their feedback on?
Line `161-166` of  [sktime/forecasting/base/adapters/_generalised_statsforecast.py](https://github.com/sktime/sktime/compare/main...Abelarm:sktime:5666_StatsForecastMSTL_with_fh?expand=1#diff-1b0922c70b453eca4eb090279c72c0c5b6eaacbcfd3b36c15c872d8e36a9db2e)

#### Did you add any tests for the change?
Yes two new tests in `sktime/forecasting/tests/test_statsforecast.py` - `test_statsforecast_mstl_with_fh` : 
- One with valid `fh` 
- One with empty `fh` that should fail in specific way...

#### Any other comments?
No

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks for identifying the issue!

Unfortunately, this introduces to the `StatsForecastBackAdapter` a coupling to the adapter, since it expects its `_fh` to be externally set.

Can we avoid this kind of coupling? Perhaps adding a new method `set_fh`?

FYI @yarnabrina, what do you think?

**Comment by Abelarm** [comment]
Let me know if you like it in this way.

File: sktime/forecasting/base/adapters/_generalised_statsforecast.py [hunk 1]
```diff
@@ -158,7 +159,11 @@ def _fit(self, y, X, fh):
         -------
         self : reference to self
         """
-        del fh  # avoid being detected as unused by ``vulture`` like tools
+        if fh is not None and hasattr(self, "_trend_forecaster"):
+            # pass the fh to _trend_forecaster in case it needs it
+            self._trend_forecaster.set_fh(fh)
+        else:
+            del fh  # avoid being detected as unused by ``vulture`` like tools
```
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_generalised_statsforecast.py]: You can safely skip these two lines. As long as `fh` is used in the function scope, it's okay.
```suggestion
```

File: sktime/forecasting/base/adapters/_generalised_statsforecast.py [hunk 2]
```diff
@@ -158,7 +159,11 @@ def _fit(self, y, X, fh):
         -------
         self : reference to self
         """
-        del fh  # avoid being detected as unused by ``vulture`` like tools
+        if fh is not None and hasattr(self, "_trend_forecaster"):
+            # pass the fh to _trend_forecaster in case it needs it
+            self._trend_forecaster.set_fh(fh)
```
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_generalised_statsforecast.py]: Can you please tell how do we know for sure `set_fh` exists? This looks like too much of a coupling to back adapter to me. Won't it be better to address from that class itself?

Full disclaimer: back adapter is written by @luca-miniati and I haven't used it myself ever and not readily familiar with the details (or whether it's still up to date with latest statsforecast mstl features). If the above not possible for some reason, at least add a check for self type. But coupling is still a concern.

File: sktime/forecasting/tests/test_statsforecast.py [hunk 3]
```diff
@@ -47,6 +47,52 @@ def test_statsforecast_mstl(mock_autoets):
     predict.assert_called_with(36, X=None, level=[95.0])
 
 
+@pytest.mark.skipif(
+    not run_test_for_class(StatsForecastMSTL),
+    reason="run test only if softdeps are present and incrementally (if requested)",
+)
+@pytest.mark.parametrize("fh", [[1, 2, 3], None], ids=["valid fh", "None fh"])
```
Comment (yarnabrina) [inline sktime/forecasting/tests/test_statsforecast.py]: Can you also add two more types?

1. scalar (`5`)
2. in-sample (`[-2, -1]`)

File: sktime/forecasting/tests/test_statsforecast.py [hunk 4]
```diff
@@ -47,6 +47,52 @@ def test_statsforecast_mstl(mock_autoets):
...
+            f"expected {model._trend_forecaster._fh} - got: {fh}"
+        )  # Assert that fh was passed correctly to trend_forecaster
+        preds = model.predict(fh=fh)
+        assert preds.shape == (len(fh),), (
+            f"Expeced the len of the prediction to be equal to the len of the forecast "
+            f"horizon passed as input, espected: ({len(fh)}, ) - got: {preds.shape}"
+        )  # Assert that prediction works as expected
+
+    except ValueError as e:
+        assert type(e) is ValueError, (
```
Comment (yarnabrina) [inline sktime/forecasting/tests/test_statsforecast.py]: If the code enters the above `except` block, it must be of type `ValueErro`. When will this `assert` ever fail?

File: sktime/forecasting/tests/test_statsforecast.py [hunk 5]
```diff
@@ -47,6 +47,52 @@ def test_statsforecast_mstl(mock_autoets):
...
+        season_length=[3, 12], trend_forecaster=reduction_forecaster
+    )
+
+    try:
+        # fit with fh passed to model
+        model.fit(y, fh=fh)
+        assert all(model._trend_forecaster._fh == fh), (
+            f"fh for the instatiated trend_forecaster is different, "
+            f"expected {model._trend_forecaster._fh} - got: {fh}"
+        )  # Assert that fh was passed correctly to trend_forecaster
```
Comment (yarnabrina) [inline sktime/forecasting/tests/test_statsforecast.py]: Probably a naive question, but `fh` is not correctly passed, what is the expected error? Can the `model.fit` still work without error if it is not passed as the value and it'll still reach this line?

File: sktime/forecasting/tests/test_statsforecast.py [hunk 6]
```diff
@@ -47,6 +47,52 @@ def test_statsforecast_mstl(mock_autoets):
...
+        preds = model.predict(fh=fh)
+        assert preds.shape == (len(fh),), (
+            f"Expeced the len of the prediction to be equal to the len of the forecast "
+            f"horizon passed as input, espected: ({len(fh)}, ) - got: {preds.shape}"
+        )  # Assert that prediction works as expected
```
Comment (yarnabrina) [inline sktime/forecasting/tests/test_statsforecast.py]: What exactly is the expectation for the size of prediction: to match the shape of `fh` passed in `fit` or the one passed in `predict`? You are passing same to both, but probably be safe to test passing separate or if that's not allowed not pass at all?

**Comment by yarnabrina** [comment]
@fkiraly if you see the CI jobs, number of tests are just 2k. This PR modified the base statsforecast adapter and we have quite a few statsforecast estimators (8-9 I think). Is the CI running as expected?

[I can't prove it's not, but the number seems too less and num skips seem too high]

File: sktime/forecasting/base/adapters/_generalised_statsforecast.py (hunk 2 above)
Comment (Abelarm) [reply sktime/forecasting/base/adapters/_generalised_statsforecast.py]: I agree I don't like it either but I am not so much familiar with the whole code.
When you say _"Won't it be better to address from that class itself?"_ what exactly do you mean?
I am open to code a more elegant solution.

**Comment by fkiraly** [comment]
> @fkiraly if you see the CI jobs, number of tests are just 2k. This PR modified the base statsforecast adapter and we have quite a few statsforecast estimators (8-9 I think). Is the CI running as expected?
> 
> [I can't prove it's not, but the number seems too less and num skips seem too high]

hm, not sure. I added an explicit condition that the child estimators should be tested if the intermediate base class changes. Let's see if the number of tests changes.

File: sktime/forecasting/tests/test_statsforecast.py (hunk 4 above)
Comment (Abelarm) [reply sktime/forecasting/tests/test_statsforecast.py]: Improved this, I actually check that `repr(e)` is exactly what I expect

File: sktime/forecasting/tests/test_statsforecast.py (hunk 5 above)
Comment (Abelarm) [reply sktime/forecasting/tests/test_statsforecast.py]: You are definitely right, gonna delete it.
Is not actually useful.

File: sktime/forecasting/base/adapters/_generalised_statsforecast.py (hunk 2 above)
Comment (Abelarm) [reply sktime/forecasting/base/adapters/_generalised_statsforecast.py]: I improved a bit (I hope), I moved the code to a method `set_fh_to_trend_forecaster` into the `_GeneralisedStatsForecastAdapter` class.
What do you think? 
There is some coupling but I don't think I can make it better (I still learning the codebase)

File: sktime/forecasting/tests/test_statsforecast.py (hunk 6 above)
Comment (Abelarm) [reply sktime/forecasting/tests/test_statsforecast.py]: I deleted this piece, my idea was to check that the prediction length was exactly the one of fh.
It's not possible to pass a different fh in `fit` and `predict` they need to be the same.

File: sktime/forecasting/tests/test_statsforecast.py (hunk 3 above)
Comment (Abelarm) [reply sktime/forecasting/tests/test_statsforecast.py]: in-sample is not supported by my test case: 
`make_reduction` of `GradientBoostingRegressor` as trend_forecaster.

If you can suggest me a estimator that requires `fh` in `fit` and does support in-sample I will happily change it.

**Comment by Abelarm** [comment]
I think I am stuck with this as well...

The BIGGEST problem is that I am not able to run the exact same test on my local machine, my vscode does does run the failing test (e.g. `test_predict_time_index_with_X`) but not for `StatsForecastMST`'s combinations so I have to tweak my way to run the test (which is time consuming and error prone)

@fkiraly or @yarnabrina  could you help me?

**Comment by yarnabrina** [comment]
I'm severely bombarded by multiple threads in work tjis week. I won't have any time to review or help before next weekend, sorry.

**Comment by Abelarm** [comment]
> I'm severely bombarded by multiple threads in work tjis week. I won't have any time to review or help before next weekend, sorry.

Problems solved, I need your feedback on the code structure (methods and coupling)

**Review by fkiraly — COMMENTED** [review]
@yarnabrina, review would be appreciated!

File: sktime/forecasting/tests/test_statsforecast.py [hunk 7]
```diff
@@ -47,6 +47,59 @@ def test_statsforecast_mstl(mock_autoets):
...
+        )
+    except NotImplementedError as e:
+        assert (
+            "DirectTabularRegressionForecaster can not perform "
+            "in-sample prediction. Found fh with in sample index:"
+        ) in repr(e), (
+            "Unexpected exception raised - should have failed with "
+            "NotImplementedError, DirectTabularRegressionForecaster "
+            "can not perform in-sample prediction ..."
+        )
```
Comment (yarnabrina) [inline sktime/forecasting/tests/test_statsforecast.py:88]: I suggest splitting this:

1. First `try` block on only `fit` where you catch only `ValueError`
2. Next `try` one on only `predict` catching `NotImplementedError`

Comment (yarnabrina) [reply sktime/forecasting/tests/test_statsforecast.py:88]: There are some estimators where `fh` is needed in `fit` and if you pass a different `fh` in `predict` it fails. Not sure whether it's worth to test that here.

File: sktime/forecasting/base/adapters/_generalised_statsforecast.py [hunk 8]
```diff
@@ -7,6 +7,7 @@
 import pandas
 
 from sktime.forecasting.base import BaseForecaster
+from sktime.forecasting.base._base import ForecastingHorizon
```
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_generalised_statsforecast.py]:
```suggestion
from sktime.forecasting.base import ForecastingHorizon
```

File: sktime/forecasting/base/adapters/_generalised_statsforecast.py [hunk 9]
```diff
@@ -128,6 +132,34 @@ def _instantiate_model(self):
...
+        """Set forecasting horizon to trend forecaster if it exists.
+
+        Parameters
+        ----------
+        fh : ForecastingHorizon or None
+            The forecasting horizon with the steps ahead to to predict.
+        y : pd.Series
+            The time series data used for fitting.
+        """
+        from statsforecast.models import MSTL
```
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_generalised_statsforecast.py]: I'm really not sure why MSTL import is needed in generalised adapter. I'm seeing this after a while, so may have lost some context, can you please elaborate?

@fkiraly what are thoughts about this additional import for whatever algorithm user plans to use from `statsforecast`?

File: sktime/forecasting/base/adapters/_generalised_statsforecast.py [hunk 10]
```diff
@@ -515,7 +574,8 @@ def predict(self, h, X=None, level=None):
             Dictionary with entries mean for point predictions and level_* for
             probabilistic predictions.
         """
-        mean = self.estimator.predict(fh=range(1, h + 1), X=X)[:, 0]
+        fh = self._get_fh() or range(1, h + 1)
```
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_generalised_statsforecast.py]: Am I reading it right that if `fh` is passed during `fit`, `h` passed during predict, which even seems to be a mandatory argument, will be simply ignored (emphasis on `if`)?

**Review by yarnabrina — CHANGES_REQUESTED** [review]
@Abelarm made some comments, and requesting for a short recap what's the problem and how are you solving it. I was away for a while, plus this is not an estimator I wrote/maintain so a bit difficult to grasp what's happening. Apologies.

File: sktime/forecasting/tests/test_statsforecast.py (hunk 7 above)
Comment (Abelarm) [reply sktime/forecasting/tests/test_statsforecast.py:88]: Actually the try is needed only for fit, in predict I don't catch anything.

File: sktime/forecasting/base/adapters/_generalised_statsforecast.py (hunk 9 above)
Comment (Abelarm) [reply sktime/forecasting/base/adapters/_generalised_statsforecast.py]: That's because I want to set the `fh` only class "extending" class is MSTL (which is the case of the bug).

If it goes against the rules of `sktime` I can leave the only these two conditions:

```python
fh is not None
and hasattr(self, "_trend_forecaster")
```
I don't know if there is another forecaster that has `_trend_forecaster` as attribute. (probably it's redundant)

File: sktime/forecasting/base/adapters/_generalised_statsforecast.py (hunk 10 above)
Comment (Abelarm) [reply sktime/forecasting/base/adapters/_generalised_statsforecast.py]: Exactly!
That's because `statsforecast` does not support the `ForecastHorizon` that `sktime` implements.
It only support a list (hence the `range(1, h+1)`)

So as you you said:
 - if `fh` is passed during fit it will utilize the one that was calculated during `fit` with its own logic.
 - Otherwise it will simply take the one passed by `predict` (which it also has it own logic but, the one written by the creator of this class)

**Comment by Abelarm** [comment]
> @Abelarm made some comments, and requesting for a short recap what's the problem and how are you solving it. I was away for a while, plus this is not an estimator I wrote/maintain so a bit difficult to grasp what's happening. Apologies.

It all started with this bug https://github.com/sktime/sktime/issues/5666. 

The bug was that `fh` passed in fit was not passed to the `trend_forecaster` of `MSTL`.

I solved it by one main function (and two helpers):
- `_set_fh_to_trend_forecaster` (which add the `fh` to the trend forecaster => `_trend_forecaster`)

The `_trend_forecaster` is a `StatsForecastBackAdapter` so I added here two helper functions:
- `set_fh`
- `_get_fh`


The complexity here was that the `MSTL` of `statsforecast` only supports list as `fh`; So a bit of logic was necessary to handle full-insample when adding to `trend_forecaster`

File: sktime/forecasting/tests/test_statsforecast.py (hunk 7 above)
Comment (Abelarm) [reply sktime/forecasting/tests/test_statsforecast.py:88]: > There are some estimators where `fh` is needed in `fit` and if you pass a different `fh` in `predict` it fails. Not sure whether it's worth to test that here.

I should fail the same, do you want me to check it as well?

File: sktime/forecasting/base/adapters/_generalised_statsforecast.py (hunk 10 above)
Comment (yarnabrina) [reply sktime/forecasting/base/adapters/_generalised_statsforecast.py]: 1. Consider a model like ARIMA. All the sktime adapters (pmdarima/statsmodels/skforecast) supports specification of fh in both fit/predict, and passing both times is also okay for overriding purposes [I really am not sure exactly what specific set of logic forces not to allow that for few estimators, but ARIMA is not one of them]. So even though the actual model supports overriding in predict, this won't support, which I'm not a fan of (but not to the point to say it's wrong/unacceptable).

2. Another thing I don't like (and this one is stronger than above) that as a user I have to pass `h` to predict as otherwise it will complain, but if I pass it may be silently ignored and overriden.

Comment (Abelarm) [reply sktime/forecasting/base/adapters/_generalised_statsforecast.py]: Let me try to give a bit more context how I understood the flow of this two classes works:

1. the `_fit` method of `_GeneralisedStatsForecastAdapter` ignores the `fh` (it actually delete it) (line 162)
    ```python
    del fh  # avoid being detected as unused by ``vulture`` like tools
    ...
    self._fitted_forecaster = self._forecaster.fit(y_fit_input, X=X_fit_input)
    ```
2. The `_predict` splits the predict into two predict one for `in_sample` a `out_sample` and calls the function `_predict_in_or_out_of_sample` (line 292)
  
3.  This function (`_predict_in_or_out_of_sample` - line 214)  has a lot of logic for handling the difference between the `sktime`'s `ForecastHorizon` and `stastforecast`'s `list horizon`.  One of this logic is this:
    ```python
    maximum_forecast_horizon = fh.to_relative(self.cutoff)[-1]
    ...
    predictions = predict_method(maximum_forecast_horizon, X=X, **level_kw)
    ```
4. It enters the `predict` of `MSTL` class that calls the `self.trend_forecaster.predict(**kwargs)` 

5. Finally it enters inside the `predict` of the class `StatsForecastBackAdapter`  (line: 500)
  that take as input:
    ```
    Parameters
    ----------
    h : int
        Forecast horizon.
    X : typing.Optional[numpy.ndarray], default=None
        Optional exogenous of shape (h, n_x).
    level : typing.Optional[typing.Tuple[int]], default=None
        Confidence levels (0-100) for prediction intervals.
    ```
    where `h` is the `maximum_forecast_horizon` calculate at point 3. Here it transforms the `h` in the format expected by `statsforecast` with:
    ```python
    mean = self.estimator.predict(fh=range(1, h + 1), X=X)[:, 0]
    ```

---

Where I acted was:

1. I get the `fh` from `fit` of `_GeneralisedStatsForecastAdapter` and call the function of `set_fh` of the `self._trend_forecaster`

2. The `_trend_forecaster` is a instance of `StatsForecastBackAdapter` and its set method just create a property `self.inner_fh` with whatever value it got from `fit` 

3. It follows the same flow of `predict` explained above

4. In the `predict` of `StatsForecastBackAdapter` which is: `self.trend_forecaster.predict(**kwargs)`
the code call the function `_get_fh` (which could return `None`)

5. The function `_get_fh` kind of does the same of what is done in point (3):
    ```python
    fh = getattr(self, "inner_fh", None) #Remember this was set in `set_fh`
    
    if fh is not None:
        if fh.is_all_in_sample():
            return fh
        else:
            maximum_forecast_horizon = fh[-1]
            return range(1, maximum_forecast_horizon + 1) # Same code of point (3)
    else:
        return None
    ```

6. The predict does the same of point (5) with the only difference that if `fh` is not null take the one that was provided by the function `_get_fh`

--- 

I hope I answer both of your points. 
This is the clearest and most detailed description I ever made HAHAHAH

Please let me know how you would have solved it ?

Comment (yarnabrina) [reply sktime/forecasting/base/adapters/_generalised_statsforecast.py]: > how you would have solved it ?

Honestly I do not know. I'd try everything I can to avoid a specific subclass specific change to be part of base class, and if I can't probably give up but won't make parent logic dependent on child logic, child should extend, not vice versa.

Can I suggest creating a private subclass of generalised statsforecast adapter with your changes, and use that with MSTL? What are your thoughts on this?

By the way, @luca-miniati is back adapter is generic or only works with MSTL? Are we expected to use it to use any sktime forecaster using Nixtla interface without needing to use it only through MSTL? @fkiraly do you know by any chance?

Comment (Abelarm) [reply sktime/forecasting/base/adapters/_generalised_statsforecast.py]: > By the way, @luca-miniati is back adapter is generic or only works with MSTL? Are we expected to use it to use any sktime forecaster using Nixtla interface without needing to use it only through MSTL? @fkiraly do you know by any chance?

How does nixtla play a role in this? isn't only for `statsforecast`? Sorry for the basic question.

By looking at this code it's only used by the `trend_forecaster` of MSTL.
`__init__` function of `StatsForecastMSTL`
```python
# checks if trend_forecaster is already wrapped with
# StatsForecastBackAdapter
if not isinstance(self._trend_forecaster, StatsForecastBackAdapter):
    # if trend_forecaster is sktime forecaster
    if isinstance(self._trend_forecaster, BaseForecaster):
        self._trend_forecaster = StatsForecastBackAdapter(
            self._trend_forecaster
        )
    else:
        raise TypeError(
            "The provided forecaster is not compatible with MSTL. Please ensure"
            " that the forecaster you pass into the model is a sktime "
            "forecaster."
          )
```

Comment (yarnabrina) [reply sktime/forecasting/base/adapters/_generalised_statsforecast.py]: > How does nixtla play a role in this? isn't only for statsforecast? Sorry for the basic question.

I meant statsforecast only. But if I'm not wrong, interface is quite similar between other Nixtla libraries, but apologies for being not specific without verification.

> By looking at this code it's only used by the trend_forecaster of MSTL.


The way I'm interpreting this is `StatsForecastBackAdapter` is not a private class (by lack of `_` at start etc.). So if it's being used or bexpected at least that users  are using this with `statsforecast` interface,  e.g. to use with `StatsForecast` class then we can't change or assume `MSTL` is the only intention. I may be wrong, please let me know what you think.

---

> Can I suggest creating a private subclass of generalised statsforecast adapter with your changes, and use that with MSTL? What are your thoughts on this?

Is this acceptable/reasonable to you?

Comment (Abelarm) [reply sktime/forecasting/base/adapters/_generalised_statsforecast.py]: Nice timing.

I just refactored the code to be more restricted only for `MSTL`. 
I moved the main function that set the `fh` of the trend forecaster inside the class `StatsForecastMSTL` so it does not affect the other forecaster of `statsforecast`. (by overriding `_fit` and `_predict`)

I decided to keep the main functionalities inside the class `StatsForecastBackAdapter` because it is actually only used as class for the `trend_forecaster` of `MSTL` (with already some coupling with `StatsForecastMSTL`)

I am quite confident the class `StatsForecastBackAdapter` could/should be private because it's it's only used by `StatsForecastMSTL` for the trend forecaster. (let me know if you want me to change it).

--- 

> > Can I suggest creating a private subclass of generalised statsforecast adapter with your changes, and use that with MSTL? What are your thoughts on this?
> 
> Is this acceptable/reasonable to you?

I thinks is not necessary anymore (maybe only transform the `StatsForecastBackAdapter` to private)

File: sktime/forecasting/base/adapters/_generalised_statsforecast.py [hunk 11]
```diff
@@ -158,8 +161,6 @@ def _fit(self, y, X, fh):
         -------
         self : reference to self
         """
-        del fh  # avoid being detected as unused by ``vulture`` like tools
-
```
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_generalised_statsforecast.py:162]: I think this change is no longer needed.

File: sktime/forecasting/base/adapters/_generalised_statsforecast.py [hunk 12]
```diff
@@ -480,6 +481,38 @@ def new(self):
...
+        Parameters
+        ----------
+        fh : ForecastingHorizon, optional
+            Forecasting horizon to set.
+
+        Returns
+        -------
+        self : returns reference to self
+        """
+        self._inner_fh = fh
```
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_generalised_statsforecast.py:498]: If it's a new attribute, I suggest to initiate it `__init__`, as recommended by almost all linters.

File: sktime/forecasting/base/adapters/_generalised_statsforecast.py [hunk 13]
```diff
@@ -480,6 +481,38 @@ def new(self):
...
+
+    def _get_MSTL_fh(self):
+        """
+        Get forecasting horizon.
+
+        Using the specifics of MSTL inherited from
+        `_predict_in_or_out_of_sample` method of `_GeneralisedStatsForecastAdapter`.
+        """
+        fh = getattr(self, "_inner_fh", None)
```
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_generalised_statsforecast.py]: If you agree to do the above, then `self._inner_fh` can be safely used directly.

File: sktime/forecasting/base/adapters/_generalised_statsforecast.py [hunk 14]
```diff
@@ -480,6 +481,38 @@ def new(self):
...
+
+        if fh is not None:
+            if fh.is_all_in_sample():
+                return fh
+            else:
+                maximum_forecast_horizon = fh[-1]
+                return range(1, maximum_forecast_horizon + 1)
+        else:
+            return None
```
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_generalised_statsforecast.py]: The `else` block is probably redundant.

File: sktime/forecasting/tests/test_statsforecast.py [hunk 15]
```diff
@@ -47,6 +47,79 @@ def test_statsforecast_mstl(mock_autoets):
...
+        model.fit(y, fh=fh)
+    except NotImplementedError as e:
+        assert (
+            "DirectTabularRegressionForecaster can not perform "
+            "in-sample prediction. Found fh with in sample index:"
+        ) in repr(e), (
+            "Unexpected exception raised - should have failed with "
+            "NotImplementedError, DirectTabularRegressionForecaster "
+            "can not perform in-sample prediction ..."
+        )
```
Comment (yarnabrina) [inline sktime/forecasting/tests/test_statsforecast.py]: Suggestion: can we assert instead on "ids" instead of text? Error match check can be kept, but just want to ensure that we are checking it fails at expected cases only.

File: sktime/forecasting/tests/test_statsforecast.py [hunk 16]
```diff
@@ -47,6 +47,79 @@ def test_statsforecast_mstl(mock_autoets):
...
+        return
+    except ValueError as e:
+        assert repr(e) == (
```
Comment (yarnabrina) [inline sktime/forecasting/tests/test_statsforecast.py]: Same as above.

File: sktime/forecasting/tests/test_statsforecast.py [hunk 17]
```diff
@@ -47,6 +47,79 @@ def test_statsforecast_mstl(mock_autoets):
...
+        )
+        return
+    if "different fh predict" in request.node.name:
+        # predict with different fh
+        fh.append(2)
+        expect_error = True
+    try:
+        model.predict(fh=fh)
+    except ValueError as e:
+        if expect_error:
```
Comment (yarnabrina) [inline sktime/forecasting/tests/test_statsforecast.py]: Can it lead to `expect_error` variable being not defined?

File: sktime/forecasting/statsforecast.py [hunk 18]
```diff
@@ -866,7 +868,7 @@ class StatsForecastMSTL(_GeneralisedStatsForecastAdapter):
         # estimator type
         # --------------
         "capability:exogenous": False,
-        "capability:pred_int": False,
+        "capability:pred_int": True,
```
Comment (yarnabrina) [inline sktime/forecasting/statsforecast.py]: Is it related to the `fh` change? If so, how?

File: sktime/forecasting/statsforecast.py
Comment (yarnabrina) [inline sktime/forecasting/statsforecast.py:895]: Does it not override above anyway?

**Review by yarnabrina — APPROVED** [review]
Approved in principle from my side given my limited understanding of the MSTL/back-adapter flow. Requesting @fkiraly  and @luca-miniati  to have a look.

@Abelarm added some comments, mostly very minimal questions/suggestions. Closed previous threads to keep it minimal.

File: sktime/forecasting/base/adapters/_generalised_statsforecast.py (hunk 14 above)
Comment (Abelarm) [reply sktime/forecasting/base/adapters/_generalised_statsforecast.py]: I removed the stratified if...
now I this is easier to read let me know what do you think

File: sktime/forecasting/statsforecast.py (hunk 18 above)
Comment (Abelarm) [reply sktime/forecasting/statsforecast.py]: Because if I run locally `check_estimator(StatsForecastMST)` I get:

```
FAILED: test_pred_int_tag[StatsForecastMSTL-0]
FAILED: test_pred_int_tag[StatsForecastMSTL-1]
FAILED: test_pred_int_tag[StatsForecastMSTL-2]
```

I am going to leave at false to see if it fails as well on CI

File: sktime/forecasting/statsforecast.py
Comment (Abelarm) [reply sktime/forecasting/statsforecast.py:895]: look like not otherwise it would not have failed locally...

**Comment by Abelarm** [comment]
> @Abelarm added some comments, mostly very minimal questions/suggestions. Closed previous threads to keep it minimal.

I implemented most of them, left some reply-comment...
Btw thx I love your attention to the details .

File: sktime/forecasting/statsforecast.py
Comment (yarnabrina) [reply sktime/forecasting/statsforecast.py:895]: It actually does override. The failure is because of a kind of stupid framework test, need to handle it better somehow. See above.

File: sktime/forecasting/statsforecast.py (hunk 18 above)
Comment (yarnabrina) [reply sktime/forecasting/statsforecast.py]: Hmm, it's because of this test, which is known to behave not ideally.

https://github.com/sktime/sktime/blob/6eabdc5127aa28476e5811587213dafc651eb50d/sktime/forecasting/tests/test_all_forecasters.py#L603

What this test does is for estimators with _predict_interval or _predict_quantiles defined, it expects them to support probabilistic forecasting, and same for the negative case. So for cases a condition is mostly true (but not always), e.g. for statsforecast or statsmodels adapters, it detects the methods even for instances andcompliuns if the tag is set false. You can safely set this to `False` (i.e. no change) and add a explicit skip of the test in `sktime/tests/_config.py` or tags (@fkiraly can suggest how, I'm not yet used to the new tag based skip format).

In fact this line is there already, update as required: https://github.com/sktime/sktime/blob/6eabdc5127aa28476e5811587213dafc651eb50d/sktime/tests/_config.py#L143

**Comment by fkiraly** [comment]
I have opened an issue on the test and would propose to merge a temporary fix that allows edge cases like this to pass:
https://github.com/sktime/sktime/issues/9033
https://github.com/sktime/sktime/pull/9034

We can then separate the discussion on what `test_pred_int_tag` should exactly test and how the extender contract should look like into #9033.

**Comment by Abelarm** [comment]
So I think this PR is good to go! 🚀

**Comment by fkiraly** [comment]
wait, why are the tests on CI passing?

**Comment by Abelarm** [comment]
> wait, why are the tests on CI passing?

?? why they shouldn't ?? HAHAHAH

**Comment by yarnabrina** [comment]
> wait, why are the tests on CI passing?

Pred int tag is being skipped in config file already, if that's what what you are wondering.

**Comment by yarnabrina** [comment]
@fkiraly is it approved from your side? Not merging as I've never used mstl myself.

**Comment by fkiraly** [comment]
> > wait, why are the tests on CI passing?
> 
> ?? why they shouldn't ?? HAHAHAH

No, I meant, why is `test_pred_int_tag` passing? I though after our discussion that it should fail.

**Comment by yarnabrina** [comment]
This is why.

https://github.com/sktime/sktime/blob/36d4b2b4fa6acc25a30d175097ec58139fa21537/sktime/tests/_config.py#L143

I posted [above](https://github.com/sktime/sktime/pull/8931#issuecomment-3478674769), not sure if you are not seeing it for some github ui issues.

(Or may be you were just responding, ignore in that case)

**Comment by Abelarm** [comment]
I can set it in the `tags` instead of the `_config.py` as part of this PR if you prefer to (even tho it will be a mixed PR)

**Comment by Abelarm** [comment]
I still don't know why this PR is stopped.

@yarnabrina accepted it!

**Comment by yarnabrina** [comment]
I did and still do, but I'd like to get a confirmation from Franz/Luca who wrote (and probably used) this estimator. You can check with Franz on Discord for a reminder, he's super active there (and I'm not).

(updating with main one more time btw)
